stablehlo/reference/Ops.cpp

```diff
@@ -550,62 +550,68 @@ Tensor evalDotGeneralOp(const Tensor &lhs, const Tensor &rhs,
+    // These are the indices that correspond to non-contracting dimensions,
+    // and they are initialized here.
     int64_t resultDim = 0;
-
-    // Indices do not change for batching dimensions.
-    for (uint64_t i = 0; i < lhsBatchingDimensions.size(); ++i, ++resultDim) {
-      lhsIdx[lhsBatchingDimensions[i]] = (*resultIt)[resultDim];
-      rhsIdx[rhsBatchingDimensions[i]] = (*resultIt)[resultDim];
+    for (size_t i = 0; i < lhsBatchingDimensions.size(); ++i, ++resultDim) {
+      lhsIndex[lhsBatchingDimensions[i]] = resultIndex[resultDim];
+      rhsIndex[rhsBatchingDimensions[i]] = resultIndex[resultDim];
     }
-
-    // The non-batching, non-contracting dimensions of the operands are copied
-    // over to the result dimensions following the batching dimensions.
-    for (uint64_t i = 0; i < lhsNonBatchingNonContractingDims.size(); i++)
-      lhsIdx[lhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
-
-    for (uint64_t i = 0; i < rhsNonBatchingNonContractingDims.size(); i++)
-      rhsIdx[rhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
-
-    auto sum = Element(resultType.getElementType(), 0.0);
-    // All pair-wise combination of contracting dimensions needs to be summed.
-    for (int64_t i = 0; i < totalContractingSize; ++i) {
-      sum = sum + lhs.get(lhsIdx) * rhs.get(rhsIdx);
-      if (contractingDimSizes.empty()) continue;
-      for (int64_t j = contractingDimSizes.size() - 1; j >= 0; --j) {
-        lhsIdx[lhsContractingDimensions[j]]++;
-        rhsIdx[rhsContractingDimensions[j]]++;
-        if (lhsIdx[lhsContractingDimensions[j]] != contractingDimSizes[j])
-          break;
-        lhsIdx[lhsContractingDimensions[j]] = 0;
-        rhsIdx[rhsContractingDimensions[j]] = 0;
+    for (size_t i = 0; i < lhsResultDims.size(); ++i, ++resultDim)
+      lhsIndex[lhsResultDims[i]] = resultIndex[resultDim];
+    for (size_t i = 0; i < rhsResultDims.size(); ++i, ++resultDim)
+      rhsIndex[rhsResultDims[i]] = resultIndex[resultDim];
+
+    // Iteration space is defined by contracting dimensions.
+    // The corresponding parts of lhsIndex and rhsIndex start at 0, 0, ..., 0.
+    // Then, we increment them lexicographically until we're out of bounds.
+    auto incrementIndices = [&]() -> LogicalResult {
+      // Implementation is heavily inspired by IndexSpaceIterator::operator++.
+      if (lhsContractingDimensions.empty()) return failure();
+      for (int64_t i = lhsContractingDimensions.size() - 1; i >= 0; --i) {
+        lhsIndex[lhsContractingDimensions[i]]++;
+        rhsIndex[rhsContractingDimensions[i]]++;
+        if (lhsIndex[lhsContractingDimensions[i]] <
+            lhs.getShape()[lhsContractingDimensions[i]])
+          return success();
+        if (i == 0) return failure();
+        lhsIndex[lhsContractingDimensions[i]] = 0;
+        rhsIndex[rhsContractingDimensions[i]] = 0;
       }
+      return success();
+    };
```

Comment:
I think that factoring out the logic to increment indices makes things easier to understand.